python/dgl/convert.py

```diff
@@ -264,6 +265,68 @@ def hetero_from_relations(rel_graphs):
         retg._edge_frames[i].update(rgrh._edge_frames[0])
     return retg
 
+def heterograph(edge_dict, num_nodes_dict=None):
+    """Create a heterogeneous graph from a dictionary between edge types and edge lists.
+
+    Parameters
+    ----------
+    edge_dict : dict
+        The dictionary between edge types and edge lists.
+
+        The edge types are specified as a triplet of (source node type name, edge type
+        name, destination node type name).
+
+        The edge lists can be either a pair of source and destination node ID lists,
+        a list of source-destination pairs, or a scipy sparse matrix.  See
+        :func:`dgl.graph` or :func:`dgl.bipartite` for examples.
+    num_nodes_dict : dict[str, int]
+        The number of nodes for each node type.
+
+        By default DGL infers the number of nodes for each node type from ``edge_dict``
+        by taking the maximum node ID plus one for each node type.
+
+    Returns
+    -------
+    DGLHeteroGraph
+    """
+    rel_graphs = []
+
+    # infer number of nodes for each node type
+    if num_nodes_dict is None:
+        num_nodes_dict = defaultdict(int)
+        for (srctype, etype, dsttype), edges in edge_dict.items():
+            if isinstance(edges, tuple):
+                nsrc = max(edges[0]) + 1
+                ndst = max(edges[1]) + 1
+            elif isinstance(edges, list):
+                src, dst = zip(*edges)
+                nsrc = max(src) + 1
+                ndst = max(dst) + 1
+            elif isinstance(edges, sp.sparse.spmatrix):
+                nsrc = edges.shape[0]
+                ndst = edges.shape[1]
+            elif isinstance(edges, nx.Graph):
+                if srctype == dsttype:
+                    nsrc = ndst = edges.number_of_nodes()
+                else:
+                    nsrc = len({n for n, d in edges.nodes(data=True) if d['bipartite'] == 0})
+                    ndst = edges.number_of_nodes() - nsrc
```

Comment:
Add an `else` branch? Also, I'd like to support DGLHeteroGraph data as well. You could create an DGLHeteroGraph in each branch and call `hetero_from_relations` at the end.

Reply:
The first loop is to infer number of nodes for each node type before actually creating DGLHeteroGraphs (and handle the last test case in `test_heterograph.test_create()`).